File: 0804(Calculator,Lms)/calculator.py
import math
import tkinter as tk
from tkinter import messagebox


class EngineeringCalculator:

  def __init__(self, root):
    self.root = root
    self.root.title("스마트 공학용 계산기")
    self.root.geometry("350x520")
    self.root.minsize(350, 520)

    self.is_engineering_mode = False
    self.clear_on_next_input = False

    self.create_widgets()
    self.setup_layout()

  def create_widgets(self):
    # 결과 화면 (Entry)
    self.entry_var = tk.StringVar(value="0")
    self.edit = tk.Entry(
        self.root,
        textvariable=self.entry_var,
        font=("Arial", 18),
        justify="right",
        state="readonly",
    )

    # 공학용 모드 토글 버튼
    self.toggle_btn = tk.Button(
        self.root,
        text="공학용 ▶",
        font=("Arial", 11, "bold"),
        command=self.toggle_mode,
    )

    # 기본 버튼 정의 (일반용 20개)
    self.buttons_info = [
        ("C", self.on_clear),
        ("BS", self.on_backspace),
        ("(", lambda: self.on_input("(", True)),
        (")", lambda: self.on_input(")", True)),
        ("7", lambda: self.on_input("7", False)),
        ("8", lambda: self.on_input("8", False)),
        ("9", lambda: self.on_input("9", False)),
        ("/", lambda: self.on_input("/", True)),
        ("4", lambda: self.on_input("4", False)),
        ("5", lambda: self.on_input("5", False)),
        ("6", lambda: self.on_input("6", False)),
        ("*", lambda: self.on_input("*", True)),
        ("1", lambda: self.on_input("1", False)),
        ("2", lambda: self.on_input("2", False)),
        ("3", lambda: self.on_input("3", False)),
        ("-", lambda: self.on_input("-", True)),
        ("0", lambda: self.on_input("0", False)),
        (".", lambda: self.on_input(".", False)),
        ("=", self.on_equals),
        ("+", lambda: self.on_input("+", True)),
    ]

    self.buttons = []
    for text, cmd in self.buttons_info:
      btn = tk.Button(
          self.root, text=text, font=("Arial", 12, "bold"), command=cmd
      )
      self.buttons.append(btn)

    # 공학용 버튼 정의 (6개)
    self.eng_buttons_info = [
        ("sin", lambda: self.on_trig("sin")),
        ("cos", lambda: self.on_trig("cos")),
        ("tan", lambda: self.on_trig("tan")),
        ("√", self.on_sqrt),
        ("x^y", lambda: self.on_input("^", True)),
        ("π", lambda: self.on_input("π", True)),
    ]

    self.eng_buttons = []
    for text, cmd in self.eng_buttons_info:
      btn = tk.Button(
          self.root, text=text, font=("Arial", 11, "bold"), command=cmd
      )
      self.eng_buttons.append(btn)

  def setup_layout(self):
    # 기존 위젯들 배치 초기화
    for btn in self.buttons:
      btn.place_forget()
    for btn in self.eng_buttons:
      btn.place_forget()

    width = self.root.winfo_width()
    height = self.root.winfo_height()

    # 창 크기가 너무 작을 때 기본값 보정
    if width < 100:
      width = 350
    if height < 100:
      height = 520

    margin = 15
    spacing = 8

    self.edit.place(
        x=margin, y=margin, width=width - (margin * 2), height=40
    )
    self.toggle_btn.place(
        x=margin, y=65, width=width - (margin * 2), height=30
    )

    top_offset = 105
    normal_available_height = height - top_offset - margin - (spacing * 4)

    if self.is_engineering_mode:
      btn_width = (width - (margin * 2) - (spacing * 5)) // 6
      btn_height = normal_available_height // 5

      # 20개 기본 버튼 (4열 x 5행)
      for i, btn in enumerate(self.buttons):
        row = i // 4
        col = i % 4
        x = margin + col * (btn_width + spacing)
        y = top_offset + row * (btn_height + spacing)
        btn.place(x=x, y=y, width=btn_width, height=btn_height)

      # 6개 공학용 버튼 (우측에 2열 너비로 세로 배치)
      eng_x = margin + 4 * (btn_width + spacing)
      eng_height = normal_available_height // 6
      for i, btn in enumerate(self.eng_buttons):
        y = top_offset + i * (eng_height + (spacing // 2))
        btn.place(
            x=eng_x,
            y=y,
            width=(btn_width * 2) + spacing,
            height=eng_height,
        )
    else:
      available_width = width - (margin * 2) - (spacing * 3)
      btn_width = available_width // 4
      btn_height = normal_available_height // 5

      # 20개 기본 버튼 (4열 x 5행)
      for i, btn in enumerate(self.buttons):
        row = i // 4
        col = i % 4
        x = margin + col * (btn_width + spacing)
        y = top_offset + row * (btn_height + spacing)
        btn.place(x=x, y=y, width=btn_width, height=btn_height)

  def toggle_mode(self):
    self.is_engineering_mode = not self.is_engineering_mode
    if self.is_engineering_mode:
      self.root.geometry("560x520")
      self.toggle_btn.config(text="일반용 ◀")
    else:
      self.root.geometry("350x520")
      self.toggle_btn.config(text="공학용 ▶")
    self.root.update_idletasks()
    self.setup_layout()

  def on_clear(self):
    self.entry_var.set("0")
    self.clear_on_next_input = False

  def on_backspace(self):
    current = self.entry_var.get()
    if len(current) > 1 and current != "Error":
      self.entry_var.set(current[:-1])
    else:
      self.entry_var.set("0")

  def on_input(self, char, allow_chain):
    current = self.entry_var.get()

    if self.clear_on_next_input:
      if allow_chain:
        self.clear_on_next_input = False
      else:
        current = ""
        self.clear_on_next_input = False

    if current == "0" or current == "Error":
      self.entry_var.set(char)
    else:
      self.entry_var.set(current + char)

  def preprocess_expression(self, raw_expr):
    # 1단계: 숫자 바로 뒤에 π가 오는 경우 * 삽입 (예: 12π -> 12*π)
    step1 = ""
    for i in range(len(raw_expr)):
      step1 += raw_expr[i]
      if i + 1 < len(raw_expr):
        if raw_expr[i].isdigit() and raw_expr[i + 1] == "π":
          step1 += "*"

    # 2단계: π를 파이썬 수학 값으로 치환
    expanded = step1.replace("π", str(math.pi))

    # 3단계: 암묵적 곱셈 처리 (숫자( or )숫자 or )( )
    processed = ""
    for i in range(len(expanded)):
      processed += expanded[i]
      if i + 1 < len(expanded):
        curr = expanded[i]
        nxt = expanded[i + 1]
        if (
            (curr.isdigit() and nxt == "(")
            or (curr == ")" and nxt.isdigit())
            or (curr == ")" and nxt == "(")
        ):
          processed += "*"

    # 파이썬 거듭제곱 연산자 처리 (^)
    processed = processed.replace("^", "**")
    return processed

  def calculate(self):
    try:
      raw_expr = self.entry_var.get()
      expr = self.preprocess_expression(raw_expr)
      # 안전한 수식 계산을 위해 eval 사용
      result = eval(expr)
      return result
    except Exception:
      return "Error"

  def on_equals(self):
    result = self.calculate()
    if result == "Error":
      self.entry_var.set("Error")
    else:
      # 정수일 경우 소수점 제거 형식 정리
      if isinstance(result, float) and result.is_integer():
        result = int(result)
      self.entry_var.set(str(result))
    self.clear_on_next_input = True

  def on_sqrt(self):
    result = self.calculate()
    if result == "Error" or result < 0:
      self.entry_var.set("Error")
    else:
      val = math.sqrt(result)
      if val.is_integer():
        val = int(val)
      self.entry_var.set(str(val))
    self.clear_on_next_input = True

  def on_trig(self, func_name):
    result = self.calculate()
    if result == "Error":
      self.entry_var.set("Error")
    else:
      rad = math.radians(result)
      if func_name == "sin":
        res = math.sin(rad)
      elif func_name == "cos":
        res = math.cos(rad)
      elif func_name == "tan":
        res = math.tan(rad)

      # 미세한 부동소수점 오차 정리 (예: tan(90) 등 방지용 보정)
      if abs(res) < 1e-12:
        res = 0.0

      self.entry_var.set(str(res))
    self.clear_on_next_input = True


if __name__ == "__main__":
  root = tk.Tk()
  app = EngineeringCalculator(root)
  root.mainloop()

File: 0804(Calculator,Lms)/calculator.cpp
#include <windows.h>
#include <string>
#include <cmath>
#include <vector>
#include <sstream>
#include <cwctype>
#include <stdexcept>

// 컨트롤 ID 정의
#define ID_EDIT 100
#define ID_BTN_TOGGLE 101

#define ID_BTN_C    0
#define ID_BTN_BS   1
#define ID_BTN_LPAREN 2
#define ID_BTN_RPAREN 3

#define ID_BTN_7    4
#define ID_BTN_8    5
#define ID_BTN_9    6
#define ID_BTN_DIV  7
#define ID_BTN_4    8
#define ID_BTN_5    9
#define ID_BTN_6    10
#define ID_BTN_MUL  11
#define ID_BTN_1    12
#define ID_BTN_2    13
#define ID_BTN_3    14
#define ID_BTN_MINUS 15
#define ID_BTN_0    16
#define ID_BTN_DOT  17
#define ID_BTN_EQ   18
#define ID_BTN_PLUS 19

#define ID_BTN_SIN 20
#define ID_BTN_COS 21
#define ID_BTN_TAN 22
#define ID_BTN_SQRT 23
#define ID_BTN_POW 24
#define ID_BTN_PI 25

// 전역 변수
HWND hEdit;
HWND hToggleBtn;
HWND hButtons[20];
HWND hEngButtons[6];

bool isEngineeringMode = false;
bool clearOnNextInput = false;

// --- [수식 계산 엔진 시작] ---
double EvaluateExpr(const std::wstring& expr, size_t& idx);

double ParseFactor(const std::wstring& expr, size_t& idx) {
    while (idx < expr.length() && expr[idx] == L' ') idx++;
    if (idx >= expr.length()) return 0;

    if (expr[idx] == L'(') {
        idx++; // '(' 스킵
        double val = EvaluateExpr(expr, idx);
        if (idx < expr.length() && expr[idx] == L')') {
            idx++; // ')' 스킵
        }
        return val;
    }

    if (expr[idx] == L'-') {
        idx++;
        return -ParseFactor(expr, idx);
    }
    if (expr[idx] == L'+') {
        idx++;
        return ParseFactor(expr, idx);
    }

    size_t startIdx = idx;
    bool hasDot = false;
    while (idx < expr.length()) {
        wchar_t c = expr[idx];
        if (iswdigit(c)) {
            idx++;
        }
        else if (c == L'.' && !hasDot) {
            hasDot = true;
            idx++;
        }
        else if (c == L'e' || c == L'E') {
            idx++;
            if (idx < expr.length() && (expr[idx] == L'+' || expr[idx] == L'-')) idx++;
        }
        else {
            break;
        }
    }

    if (startIdx == idx) return 0;

    std::wstring numStr = expr.substr(startIdx, idx - startIdx);
    return _wtof(numStr.c_str());
}

double ParseTerm(const std::wstring& expr, size_t& idx) {
    double val = ParseFactor(expr, idx);
    while (idx < expr.length()) {
        while (idx < expr.length() && expr[idx] == L' ') idx++;
        if (idx < expr.length() && expr[idx] == L'*') {
            idx++;
            val *= ParseFactor(expr, idx);
        }
        else if (idx < expr.length() && expr[idx] == L'/') {
            idx++;
            double divisor = ParseFactor(expr, idx);
            if (divisor == 0) throw std::runtime_error("Division by zero");
            val /= divisor;
        }
        else if (idx < expr.length() && expr[idx] == L'^') {
            idx++;
            val = pow(val, ParseFactor(expr, idx));
        }
        else {
            break;
        }
    }
    return val;
}

double EvaluateExpr(const std::wstring& expr, size_t& idx) {
    double val = ParseTerm(expr, idx);
    while (idx < expr.length()) {
        while (idx < expr.length() && expr[idx] == L' ') idx++;
        if (idx < expr.length() && expr[idx] == L'+') {
            idx++;
            val += ParseTerm(expr, idx);
        }
        else if (idx < expr.length() && expr[idx] == L'-') {
            idx++;
            val -= ParseTerm(expr, idx);
        }
        else {
            break;
        }
    }
    return val;
}

// 전처리: 암묵적 곱셈 및 π 처리
std::wstring PreprocessExpression(const std::wstring& rawExpr) {
    // 1단계: 숫자 바로 뒤에 π가 오는 경우 사이에 * 추가 (예: 12π -> 12*π)
    std::wstring step1 = L"";
    for (size_t i = 0; i < rawExpr.length(); ++i) {
        step1 += rawExpr[i];
        if (i + 1 < rawExpr.length()) {
            wchar_t curr = rawExpr[i];
            wchar_t next = rawExpr[i + 1];
            if (iswdigit(curr) && next == L'π') {
                step1 += L'*';
            }
        }
    }

    // 2단계: 'π'를 실제 값으로 치환
    std::wstring expanded = L"";
    for (size_t i = 0; i < step1.length(); ++i) {
        if (step1[i] == L'π') {
            expanded += L"3.141592653589793";
        }
        else {
            expanded += step1[i];
        }
    }

    // 3단계: 괄호 및 숫자 간의 암묵적 곱셈(*) 자동 삽입 (2(5), (2)5, (2)(3) 등)
    std::wstring processed = L"";
    for (size_t i = 0; i < expanded.length(); ++i) {
        processed += expanded[i];
        if (i + 1 < expanded.length()) {
            wchar_t curr = expanded[i];
            wchar_t next = expanded[i + 1];

            if ((iswdigit(curr) && next == L'(') ||
                (curr == L')' && iswdigit(next)) ||
                (curr == L')' && next == L'(')) {
                processed += L'*';
            }
        }
    }
    return processed;
}

double CalculateExpression(const wchar_t* wExpr) {
    try {
        std::wstring rawExpr(wExpr);
        std::wstring expr = PreprocessExpression(rawExpr);
        size_t idx = 0;
        return EvaluateExpr(expr, idx);
    }
    catch (...) {
        return 0;
    }
}
// --- [수식 계산 엔진 끝] ---

// 레이아웃 배치 함수
void UpdateLayout(HWND hwnd) {
    RECT rect;
    GetClientRect(hwnd, &rect);
    int width = rect.right - rect.left;
    int height = rect.bottom - rect.top;

    int margin = 15;
    int spacing = 8;

    MoveWindow(hEdit, margin, margin, width - (margin * 2), 40, TRUE);
    MoveWindow(hToggleBtn, margin, 65, width - (margin * 2), 30, TRUE);

    int topOffset = 105;
    int availableWidth = width - (margin * 2) - (spacing * 3);
    int normalAvailableHeight = height - topOffset - margin - (spacing * 4);

    if (isEngineeringMode) {
        int btnWidth = (width - (margin * 2) - (spacing * 5)) / 6;
        int btnHeight = normalAvailableHeight / 5;

        for (int i = 0; i < 20; ++i) {
            int row = i / 4;
            int col = i % 4;
            int x = margin + col * (btnWidth + spacing);
            int y = topOffset + row * (btnHeight + spacing);
            MoveWindow(hButtons[i], x, y, btnWidth, btnHeight, TRUE);
            ShowWindow(hButtons[i], SW_SHOW);
        }

        int engX = margin + 4 * (btnWidth + spacing);
        int engHeight = normalAvailableHeight / 6;
        for (int i = 0; i < 6; ++i) {
            int y = topOffset + i * (engHeight + (spacing / 2));
            MoveWindow(hEngButtons[i], engX, y, (btnWidth * 2) + spacing, engHeight, TRUE);
            ShowWindow(hEngButtons[i], SW_SHOW);
        }
    }
    else {
        int btnWidth = availableWidth / 4;
        int btnHeight = normalAvailableHeight / 5;

        for (int i = 0; i < 20; ++i) {
            int row = i / 4;
            int col = i % 4;
            int x = margin + col * (btnWidth + spacing);
            int y = topOffset + row * (btnHeight + spacing);
            MoveWindow(hButtons[i], x, y, btnWidth, btnHeight, TRUE);
            ShowWindow(hButtons[i], SW_SHOW);
        }

        for (int i = 0; i < 6; ++i) {
            ShowWindow(hEngButtons[i], SW_HIDE);
        }
    }
}

// 윈도우 프로시저
LRESULT CALLBACK WndProc(HWND hwnd, UINT msg, WPARAM wp, LPARAM lp) {
    switch (msg) {
    case WM_CREATE: {
        hEdit = CreateWindowW(L"EDIT", L"0",
            WS_CHILD | WS_VISIBLE | WS_BORDER | ES_RIGHT | ES_READONLY,
            15, 15, 310, 40, hwnd, (HMENU)ID_EDIT, NULL, NULL);

        hToggleBtn = CreateWindowW(L"BUTTON", L"공학용 ▶",
            WS_CHILD | WS_VISIBLE | BS_PUSHBUTTON,
            15, 65, 310, 30, hwnd, (HMENU)ID_BTN_TOGGLE, NULL, NULL);

        const wchar_t* btnLabels[20] = {
            L"C",  L"BS", L"(",  L")",
            L"7",  L"8",  L"9",  L"/",
            L"4",  L"5",  L"6",  L"*",
            L"1",  L"2",  L"3",  L"-",
            L"0",  L".",  L"=",  L"+"
        };
        int ids[20] = {
            ID_BTN_C, ID_BTN_BS, ID_BTN_LPAREN, ID_BTN_RPAREN,
            ID_BTN_7, ID_BTN_8, ID_BTN_9, ID_BTN_DIV,
            ID_BTN_4, ID_BTN_5, ID_BTN_6, ID_BTN_MUL,
            ID_BTN_1, ID_BTN_2, ID_BTN_3, ID_BTN_MINUS,
            ID_BTN_0, ID_BTN_DOT, ID_BTN_EQ, ID_BTN_PLUS
        };
        for (int i = 0; i < 20; ++i) {
            hButtons[i] = CreateWindowW(L"BUTTON", btnLabels[i],
                WS_CHILD | WS_VISIBLE | BS_PUSHBUTTON,
                0, 0, 0, 0, hwnd, (HMENU)(INT_PTR)ids[i], NULL, NULL);
        }

        const wchar_t* engLabels[6] = { L"sin", L"cos", L"tan", L"√", L"x^y", L"π" };
        int engIds[6] = { ID_BTN_SIN, ID_BTN_COS, ID_BTN_TAN, ID_BTN_SQRT, ID_BTN_POW, ID_BTN_PI };
        for (int i = 0; i < 6; ++i) {
            hEngButtons[i] = CreateWindowW(L"BUTTON", engLabels[i],
                WS_CHILD | WS_VISIBLE | BS_PUSHBUTTON,
                0, 0, 0, 0, hwnd, (HMENU)(INT_PTR)engIds[i], NULL, NULL);
            ShowWindow(hEngButtons[i], SW_HIDE);
        }
        break;
    }
    case WM_SIZE: {
        UpdateLayout(hwnd);
        break;
    }
    case WM_COMMAND: {
        int id = LOWORD(wp);
        wchar_t buf[256];

        if (id == ID_BTN_TOGGLE) {
            isEngineeringMode = !isEngineeringMode;
            RECT rc;
            GetWindowRect(hwnd, &rc);
            int newWidth = isEngineeringMode ? 560 : 350;
            SetWindowPos(hwnd, NULL, rc.left, rc.top, newWidth, rc.bottom - rc.top, SWP_NOMOVE | SWP_NOZORDER);

            if (isEngineeringMode) {
                SetWindowTextW(hToggleBtn, L"일반용 ◀");
            }
            else {
                SetWindowTextW(hToggleBtn, L"공학용 ▶");
            }
            UpdateLayout(hwnd);
        }
        else if (id == ID_BTN_C) {
            SetWindowTextW(hEdit, L"0");
            clearOnNextInput = false;
        }
        else if (id == ID_BTN_BS) {
            GetWindowTextW(hEdit, buf, 256);
            int len = wcslen(buf);
            if (len > 1 && wcscmp(buf, L"Error") != 0) {
                buf[len - 1] = L'\0';
                SetWindowTextW(hEdit, buf);
            }
            else {
                SetWindowTextW(hEdit, L"0");
            }
        }
        else {
            const wchar_t* inputStr = NULL;
            bool allowChain = false;

            if (id >= ID_BTN_7 && id <= ID_BTN_9) {
                wchar_t temp[2] = { (wchar_t)(L'7' + (id - ID_BTN_7)), L'\0' };
                inputStr = temp;
            }
            else if (id >= ID_BTN_4 && id <= ID_BTN_6) {
                wchar_t temp[2] = { (wchar_t)(L'4' + (id - ID_BTN_4)), L'\0' };
                inputStr = temp;
            }
            else if (id >= ID_BTN_1 && id <= ID_BTN_3) {
                wchar_t temp[2] = { (wchar_t)(L'1' + (id - ID_BTN_1)), L'\0' };
                inputStr = temp;
            }
            else if (id == ID_BTN_0) {
                inputStr = L"0";
            }
            else if (id == ID_BTN_DOT) {
                inputStr = L".";
            }
            else if (id == ID_BTN_LPAREN) {
                inputStr = L"("; allowChain = true;
            }
            else if (id == ID_BTN_RPAREN) {
                inputStr = L")"; allowChain = true;
            }
            else if (id == ID_BTN_PLUS) {
                inputStr = L"+"; allowChain = true;
            }
            else if (id == ID_BTN_MINUS) {
                inputStr = L"-"; allowChain = true;
            }
            else if (id == ID_BTN_MUL) {
                inputStr = L"*"; allowChain = true;
            }
            else if (id == ID_BTN_DIV) {
                inputStr = L"/"; allowChain = true;
            }
            else if (id == ID_BTN_POW) {
                inputStr = L"^"; allowChain = true;
            }
            else if (id == ID_BTN_PI) {
                inputStr = L"π"; allowChain = true;
            }

            if (inputStr != NULL) {
                GetWindowTextW(hEdit, buf, 256);

                if (clearOnNextInput) {
                    if (allowChain) {
                        clearOnNextInput = false;
                    }
                    else {
                        SetWindowTextW(hEdit, L"");
                        buf[0] = L'\0';
                        clearOnNextInput = false;
                    }
                }

                if (wcscmp(buf, L"0") == 0 || wcscmp(buf, L"Error") == 0) {
                    SetWindowTextW(hEdit, inputStr);
                }
                else {
                    wcscat_s(buf, 256, inputStr);
                    SetWindowTextW(hEdit, buf);
                }
            }
            else if (id == ID_BTN_EQ) {
                GetWindowTextW(hEdit, buf, 256);
                double result = CalculateExpression(buf);
                swprintf_s(buf, L"%g", result);
                SetWindowTextW(hEdit, buf);
                clearOnNextInput = true;
            }
            else if (id == ID_BTN_SQRT) {
                GetWindowTextW(hEdit, buf, 256);
                double val = CalculateExpression(buf);
                if (val < 0) SetWindowTextW(hEdit, L"Error");
                else {
                    swprintf_s(buf, L"%g", sqrt(val));
                    SetWindowTextW(hEdit, buf);
                }
                clearOnNextInput = true;
            }
            else if (id == ID_BTN_SIN || id == ID_BTN_COS || id == ID_BTN_TAN) {
                GetWindowTextW(hEdit, buf, 256);
                double val = CalculateExpression(buf);
                double rad = val * 3.141592653589793 / 180.0;
                double res = 0;
                if (id == ID_BTN_SIN) res = sin(rad);
                else if (id == ID_BTN_COS) res = cos(rad);
                else if (id == ID_BTN_TAN) res = tan(rad);

                swprintf_s(buf, L"%g", res);
                SetWindowTextW(hEdit, buf);
                clearOnNextInput = true;
            }
        }
        break;
    }
    case WM_DESTROY:
        PostQuitMessage(0);
        break;
    default:
        return DefWindowProcW(hwnd, msg, wp, lp);
    }
    return 0;
}

int WINAPI WinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance, LPSTR lpCmdLine, int nCmdShow) {
    WNDCLASSW wc = { 0 };
    wc.lpfnWndProc = WndProc;
    wc.hInstance = hInstance;
    wc.hCursor = LoadCursor(NULL, IDC_ARROW);
    wc.hbrBackground = (HBRUSH)(COLOR_BTNFACE + 1);
    wc.lpszClassName = L"EngCalculatorClass";

    RegisterClassW(&wc);

    HWND hwnd = CreateWindowW(L"EngCalculatorClass", L"스마트 공학용 계산기",
        WS_OVERLAPPEDWINDOW, CW_USEDEFAULT, CW_USEDEFAULT, 350, 520,
        NULL, NULL, hInstance, NULL);

    ShowWindow(hwnd, nCmdShow);
    UpdateWindow(hwnd);

    MSG msg;
    while (GetMessageW(&msg, NULL, 0, 0)) {
        TranslateMessage(&msg);
        DispatchMessageW(&msg);
    }

    return (int)msg.wParam;
}